File: backend/src/routes/sponsor.ts
import express, { Router } from "express";
import got, { Got } from "got";
import { logger } from "../utils/logger.js";

export interface Sponsor {
    id: number;
    name: string;
    imageUrl: string;
    url: string;
    tone: "gold" | "silver" | "bronze";
}

let sponsorCache: Sponsor[] | null = null;

export function setupSponsorRoutes(app: express.Application) {
    const router: Router = express.Router();
    const gotClient: Got = got.extend({
        prefixUrl: "https://galaxy.tianpao.top/",
        headers: {
            "User-Agent": "DeEarthX",
        },
        responseType: "json",
    });

    // GET /sponsor/ - 获取所有赞助商（公开API）
    router.get("/", async (_req, res) => {
        if (sponsorCache) {
            logger.info("返回缓存的赞助商列表", { count: sponsorCache.length });
            res.json(sponsorCache);
            return;
        }

        try {
            const response = await gotClient.get<Sponsor[]>("sponsor/");
            sponsorCache = response.body;
            logger.info("获取赞助商列表成功", { count: response.body.length });
            res.json(response.body);
        } catch (error) {
            logger.error("获取赞助商列表失败", error as Error);
            res.status(500).json({ error: "获取赞助商列表失败" });
        }
    });

    app.use("/sponsor", router);
}

File: core/sponsor/sponsor.go
package sponsor

import (
	"encoding/json"
	"sync"

	"resty.dev/v3"

	"deearthx/core/util"
)

// Sponsor represents a sponsor entry
type Sponsor struct {
	ID       int    `json:"id"`
	Name     string `json:"name"`
	ImageURL string `json:"imageUrl"`
	URL      string `json:"url"`
	Tone     string `json:"tone"` // "gold" | "silver" | "bronze"
}

// SponsorService handles sponsor operations
type SponsorService struct {
	client      *resty.Client
	cache       []Sponsor
	cacheMutex  sync.RWMutex
}

// NewSponsorService creates a new SponsorService
func NewSponsorService() *SponsorService {
	return &SponsorService{
		client: resty.New().
			SetBaseURL("https://galaxy.tianpao.top/").
			SetHeader("User-Agent", "DeEarthX"),
	}
}

// List returns all sponsors
func (ss *SponsorService) List() ([]Sponsor, error) {
	// Check cache first
	ss.cacheMutex.RLock()
	if ss.cache != nil {
		ss.cacheMutex.RUnlock()
		util.Logger.Debug("Returning cached sponsors", "count", len(ss.cache))
		return ss.cache, nil
	}
	ss.cacheMutex.RUnlock()

	// Fetch from API
	resp, err := ss.client.R().Get("sponsor/")
	if err != nil {
		util.Logger.Error("Failed to get sponsors: " + err.Error())
		return nil, err
	}

	if resp.StatusCode() >= 400 {
		util.Logger.Error("Sponsor API error: HTTP " + resp.Status())
		return nil, nil
	}

	contentType := resp.Header().Get("Content-Type")
	if contentType != "" && len(contentType) >= 16 && contentType[:16] != "application/json" {
		util.Logger.Warn("Sponsor API returned non-JSON response", "contentType", contentType)
		return nil, nil
	}

	var sponsors []Sponsor
	if err := json.Unmarshal(resp.Bytes(), &sponsors); err != nil {
		util.Logger.Warn("Failed to parse sponsor response", "error", err.Error())
		return nil, nil
	}

	// Cache result
	ss.cacheMutex.Lock()
	ss.cache = sponsors
	ss.cacheMutex.Unlock()

	util.Logger.Info("Fetched sponsors", "count", len(sponsors))
	return sponsors, nil
}

// ClearCache clears the sponsor cache
func (ss *SponsorService) ClearCache() {
	ss.cacheMutex.Lock()
	ss.cache = nil
	ss.cacheMutex.Unlock()
}
